### PR TITLE
Add configuration of check_shapes.

### PR DESCRIPTION
**Pull request by jesnie**
Add configuration of check_shapes.
1. Allow people to disable shape checking.
2. Allow people to set the format of docstrings, though right now we only support "Sphinx" and "None".

### CODE REVIEW

File: gpflow/experimental/check_shapes/__init__.py [hunk 1]
```diff
@@ -192,6 +206,11 @@ def linear_model(
         Model predictions.
     \"\"\"
 
+if you do not wish to have your docstrings rewritten, you can disable it with
+:func:`set_rewrite_docstrings`::
+
+    set_rewrite_docstrings(None)
```
Comment (vdutor) [inline gpflow/experimental/check_shapes/__init__.py:212]: Can this be made to have a boolean input similar to `set_enable_check_shapes`?

File: gpflow/experimental/check_shapes/config.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    Rewrite docstrings in the `Sphinx <https://www.sphinx-doc.org/en/master/>`_ format.
+    """
+
+    NONE = "none"
+    """
+    Do not rewrite docstrings.
+    """
+
+
+_docstring_format = DocstringFormat.SPHINX
```
Comment (vdutor) [inline gpflow/experimental/check_shapes/config.py:36]: Similarly as above, would it make sense to also specify this as a [gpflow config](https://github.com/GPflow/GPflow/blob/develop/gpflow/config/__config__.py)?

File: tests/gpflow/experimental/check_shapes/test_check_shapes.py [hunk 3]
```diff
@@ -342,6 +343,39 @@ def f(x: Optional[Input]) -> Tuple[Output, ...]:
...
+        "return: [d...]",
+    )
+    def g(a: TensorType, b: TensorType) -> TensorType:
+        return a
+
+    with pytest.raises(ShapeMismatchError):
+        g(t(2, 3), t(2, 4))
+
+    with disable_check_shapes():
+        f(t(2, 3), t(2, 4))  # Wrong shape, but checks disabled.
```
Comment (vdutor) [inline tests/gpflow/experimental/check_shapes/test_check_shapes.py:373]:
```suggestion
        g(t(2, 3), t(2, 4))  # Wrong shape, but checks disabled.
```

Comment (vdutor) [reply tests/gpflow/experimental/check_shapes/test_check_shapes.py:373]: Would this pass for `g`?

File: tests/gpflow/experimental/check_shapes/test_inheritance.py [hunk 4]
```diff
@@ -180,6 +183,50 @@ def f(self, a: TensorType) -> TensorType:
...
+            def f(self, a: TensorType, b: TensorType) -> TensorType:
+                return a
+
+    sub = SubClass()
+
+    with call_context():
+        sub.f(t(2, 3), t(2, 4))  # Wrong shape, checks maybe disabled.
+
+    with disable_check_shapes():
+        sub.f(t(2, 3), t(2, 4))  # Wrong shape, but checks disable.
```
Comment (vdutor) [inline tests/gpflow/experimental/check_shapes/test_inheritance.py:227]:
```suggestion
        sub.f(t(2, 3), t(2, 4))  # Wrong shape, but checks disabled.
```

File: tests/gpflow/experimental/check_shapes/test_inheritance.py [hunk 5]
```diff
@@ -180,6 +183,50 @@ def f(self, a: TensorType) -> TensorType:
                 return t(1)
 
 
+@pytest.mark.parametrize("enable_super", [True, False])
+@pytest.mark.parametrize("enable_sub", [True, False])
+def test_inherit_check_shapes__disable(enable_super: bool, enable_sub: bool) -> None:
```
Comment (vdutor) [inline tests/gpflow/experimental/check_shapes/test_inheritance.py:188]: Could you add a docstring that explains what you want to test here?

**Review by vdutor — APPROVED** [review]

File: tests/gpflow/experimental/check_shapes/test_inheritance.py (hunk 5 above)
Comment (jesnie) [reply tests/gpflow/experimental/check_shapes/test_inheritance.py:188]: > Could you add a docstring that explains what you want to test here?

Sure.

File: tests/gpflow/experimental/check_shapes/test_check_shapes.py (hunk 3 above)
Comment (jesnie) [reply tests/gpflow/experimental/check_shapes/test_check_shapes.py]: > Would this pass for `g`?

You're right. This should be `g`.

**Pull request merged**